Functionality upgrade due to cdo and nco version update, added scripts for direct GCM data download

yoselita opened the pull request:
This PR includes multiple changes. These are the major ones:
1. Multiple cdo and nco command changes due to the update of the software version (cdo from v1.6.3 to v1.9.8; nco  from 4.4.4 to 4.9.4)
2. Enabled interpolation of the data to desired frequency by introducing variable interval in BCtable (e.g. to 3 hourly, 6 hourly) 
3. Added a function that interpolates  from M-hourly to N-hourly frequency added (M>N).
4. Fixed bug affecting ungrib.exe (https://github.com/Yoselita/bctool/issues/2) 
5. Changed removing Feb 29 - the preprocessor.ESGF reads calendar from the data and remove_leap from BCtable, and if calendar is "standard" and remove_leap "true", the day will be removed.
6. Added scripts for GCM model data download.
7. Updated README files accordingly.

yoselita committed:
Temporal interpolation of fixed files included. For SST tos instead of tslsi.

yoselita committed:
nco and cdo commands adjusted to work correctly with the newer versions of cdo (tested with 1.9.8) and nco (tested with 4.9.4).

yoselita committed:
Height of the surface variables set to 0, to be included at the first vertical level.

yoselita committed:
An option to choose interpolation interval in time included in the header (interval). Names for temporal interpolation filters changed.

yoselita committed:
Ungrib runs only 1 time as fixed and soil variabels are to be interpolated at the requested frequency. Reading and the function for updating namelists changed - includes option to read different intial and ending times.

yoselita committed:
SM and ST at the deepest layer for the CanESM2 model included (40-405m). Interpolation techinques added for SST inteprolation.

yoselita committed:
Order of the functions changes. Removing of the leap day set (used the calendar metadata) in the function time_slice and removed from the other functions. Vertical interpolation adjusted to the newer version of cdo and nco (tested for cdo 1.9.8 nco 4.9.4). Readign and creating parameters needed for interpolating from model to pressure levels fixed. Hardcoded parameters removed when dealing with soil variables - set to be able to read soil data from models other than CanESM2. Temporal interpolation not fixed to 6 hours any more - can be swithced to 3 hours - reads the interval from the header of the BC table.

yoselita committed:
Vtable for CanESM2 model adjusted to be able to read the soil variabels at the lowest soil layer when running ungrib.exe.

yoselita committed:
Reading fixed files in the metgrid section commented.

yoselita committed:
Updated according to recent changes.

jesusff committed:
Update run_test.sh

yoselita committed:
Added filter, changed header

3D data on ESGF usually available on 6hr frequency. To create 3hr boundary data temoral interpolation needed.
Filter for interpolation of 6 hourly 3D data (e.g. 6hr to 3 hr) added.
Header adjusted for a specific test case, and variable "interval", denoting the frequecy of the output boundary data, added.

yoselita committed:
Adjusted for testing historical MPIESM12LR run

yoselita committed:
Add temporal interpolation function; add -s to several cdo commands; update vertical interpolation function

A function "Mh_to_Nh" added to interpolate Mhourly data to Nhourly data (M > N) using the cdo command "inttime" - useful if 6hourly data are available, and 3hourly output frequency is needed.
"-s" added to all cdo command to more have more clean output in the terminal/logfile while running the script (just a cosmetic change).
In the function that sets the hybrid levels "set_hybrid_levels" ap_bnds and b_bnds are kept in the output file, since cdo v1.9.8 complains if interpolation or masking is performed afterwards.

yoselita committed:
uncomment lines

yoselita committed:
Change filter for temporal interpolation of 6hourly variables

yoselita committed:
Merge pull request #1 from Yoselita/jesusff-patch-1

Update run_test.sh

yoselita committed:
Merge with Yoselita/test-patch

zequihg50 committed:
Added scripts for data download

yoselita committed:
Merge pull request #3 from zequihg50/devel

Added scripts for data download

yoselita committed:
Removed filter for fixed variables, added "remove_leap" to the header

It is not needed to interpolate in time fixed fields to run WPS without multiple runs of ungrib.exe.
"remove leap" added to the header so the preprocessor.ESGF can recognise if removing 29 Feb is desired (true) or not (false).

yoselita committed:
Change from relative to absolute time in the output grib files

ungrib.exe is not able to read correctly realtive timestamp. Setting time units to "hours" (cdo -settunits,hours) fixes the problem.
Variable noleap is read from a BCtable. This variabels controls if the leap year will be removed or not:
For the evaluation runs typically is is not removed (false), and for the historical and scenario runs it is (true) in case the GCM uses "standard" calendar.

yoselita committed:
Merge branch 'test-patch' into devel

Solves the issue #2

yoselita committed:
Update README_BCtable.md

Added explanation of header variables and a function.

yoselita committed:
Update README_BCtable.md

yoselita committed:
Added "remove_leap false" to the header

yoselita committed:
Changed and added filters, extended header

"interval" and "remove_leap" added added to the headers. Time interpolation filters renamed, and added to the surface variables

jesusff commented on README_BCtable.md:
```diff
@@ -96,10 +104,16 @@ output stream to be used by the next filter. This is a list of available filters
 
-**day_to_6h**
-:  Temporal interpolation in time from daily to 6-hourly frequency. Recommended
+**Mh_to_Nh**
```
This filter name is quite cryptic. Maybe `time_interp_to_interval`?

Yoselita replied:
I agree. time_interp_to_interval looks much better.

jesusff commented on README_BCtable.md:
```diff
@@ -96,10 +104,16 @@ output stream to be used by the next filter. This is a list of available filters
+**Mh_to_Nh**
+:  Interpolation in time from M-hourly frequency to N-hourly frequency (M > N).
+
+**day_to_Nh**
```
Given that you defined the `interval` variable in the BCtable, we could call this filter `day_to_interval`, and the same for other `Nh` filters

Yoselita replied:
Yes. I am testing at the moment if these 2 filters (Mh_to_Nh and  day_to_Nh) can be merged into one. If so, the name "time_interp_to_interval" will be set as the name of the filter.

jesusff commented on run_test.sh:
```diff
@@ -26,6 +28,30 @@ case ${model} in
+
+
+# Functions for updating namelists
+function format_date(){
```
What is the purpose of this new function? The logic behind `set_nml` below has became quite more complex

Yoselita replied:
To separate sformateddate and eformateddate into year, month, day, and hour to correctly update start time and end time in namelist.wps and namlist.input. In the original version I had problems when testing WRF and WPS because dates in namelist.wps and namelist.input were taken directly from the template.

jesusff replied:
Ok. But this function relies heavily on global variable definitions for input and output. It would be simpler to replace the call below (L75) by:
```bash
read year month day hour trash <<< `echo ${sformateddate} | tr '_T:-' '    '`
``` 
or similar (you'll still need to switch between s/eformatteddate)

jesusff reviewed (changes requested):
Thanks Josipa, I left some comments in the code. Please, also change the title to better reflect the purpose of the PR. It is quite generic.

jesusff commented:
Regarding point 4. Could you test the `cdo -a` flag (absolute time axis) to see whether it solves the issue without tampering with the units?

yoselita committed:
Added function that controls leap years

Removing a leap year, if required, solved through the function "unleap". The function is included in the wrfpostfilter as the last one.

yoselita committed:
Renamed interpolation filters

yoselita committed:
Merged with the branch -test-patch

yoselita committed:
Described new unleap filter; Renamed, removed, updated interpolation filters.

yoselita committed:
Removed "set -u" check

This option affects variables. If variable not defined/used, error is produced, which causes the program to immediately exit.
For example, if a variable that represents land_mask (sftlf) is not processed, the script will exit producing an error "land_mask: unbound variable".
To avoid this, "set -u" has been removed.

yoselita committed:
Merge branch 'test-patch' into devel

yoselita committed:
Added one line at the end of the BC tables

preprocessor.ESGF was exiting with failure since "while read" loop that was reading BC table was not completed correctly. Ending the BC tables with 1 line starting with "#" asures that the loop completes, and that preprocessor.ESGF exits with success if all the rest done correctly.

yoselita committed:
Added num_metgrid_soil_levels, removed unnecessary

yoselita committed:
Updated: Removed function format_date, comments added, reorganised

Automatic updating of the namelist.wps and namelist.input with dates corresponding to the forcing dates better formatted.
Explanation comments added.

yoselita committed:
Changed psl interpolation filter

Interpolation filter for psl changed from fixed2interval to tinterp2interval. fixed2interval keeps only the first and the last timestep,
and interpolates the steps between. Therefore it is appropriate only for the fixed variables and those that are used only for the initilazation such as soil moisture and soil temperature.
tinterp2interval keeps all the exixting timesteps, and interpolates those that are missing.

yoselita committed:
Updated function unleap

The script was aborting execution on error if 29 Feb was not in the file when applying unleap filter. To avoid this, dealing with 29 Feb changed.

jesusff commented:
OK, this is a lot of changes. I'll merge them into `devel` and then tackle any remaining detail in smaller, focused issues
